src/main/java/org/osiam/storage/query/UserFilterExpression.java

```diff
@@ -0,0 +1,87 @@
+package org.osiam.storage.query;
+
+import org.osiam.resources.scim.User;
+import org.osiam.storage.entities.UserEntity;
+import org.slf4j.LoggerFactory;
+
+import java.util.Locale;
+
+public class UserFilterExpression extends FilterExpression<UserEntity> {
+
+    private Field field;
+
+
+    UserFilterExpression(String field, FilterConstraint constraint, String value) {
+        super(constraint, value);
+        this.field = new Field(field);
+    }
+
+    @Override
+    public UserFilterExpression.Field getField() {
+        return field;
+    }
+
+    public static class Field implements FilterExpression.Field<UserEntity> {
+        private static final org.slf4j.Logger LOG = LoggerFactory.getLogger(UserFilterExpression.Field.class);
+        private String urn;
+        private String fieldName;
+        private QueryField queryField;
+
+        public Field(String field) {
+            if (field.startsWith(User.SCHEMA)) {
+                fieldName = field.substring(User.SCHEMA.length() + 1).toLowerCase(Locale.ENGLISH);
+
+            } else {
+                fieldName = field.toLowerCase(Locale.ENGLISH);
+            }
+
+            queryField = UserQueryField.fromString(fieldName);
+
+            if (queryField == null) {
+
+                int lastIndexOfPeriod = field.lastIndexOf('.');
+                int lastIndexOfColon = field.lastIndexOf(':');
+
+                if (lastIndexOfPeriod > lastIndexOfColon) {
+                    // separated by colon, log a warning!
+                    splitString(field, lastIndexOfPeriod);
+                    LOG.warn(
+                            String.format("Period (.) used as field separator in %s. " +
+                                    "This is not SCIM conform and deprecated!", field));
+                } else if (lastIndexOfColon > lastIndexOfPeriod) {
+                    splitString(field, lastIndexOfColon);
+                } else {
+                    throw new IllegalArgumentException(String.format("Unable to parse field or extension from %s",
+                            field));
+                }
+            } else {
+                urn = User.SCHEMA;
+            }
+        }
+
+        @Override
+        public String getUrn() {
+            return urn;
+        }
+
+        @Override
+        public String getFieldName() {
+            return fieldName;
+        }
+
+        @Override
+        public QueryField getQueryField() {
+            return queryField;
+        }
+
+        @Override
+        public boolean isExtension() {
+            return queryField == null;
+        }
+
+        private void splitString(String urnAndField, int lastIndexOf) {
```

Comment:
I don't like this extracted method. It's not obvious that it has a side effect and therefore the name is misleading. It's also not obvious what the process in the constructor does. I'd suggest to inline this method.